### ISSUE TITLE
about GC matrix

### ISSUE THREAD

**Issue by reherese** [icc2115/Neural-GC#17]
dear devleper：
       thank you for your share，I'm very interested in your work. Currently, I'm trying to reproduce your code, and I've made some modifications. My goal is to display the impact of each variable on the others, such as in a 6x6 matrix. I'll share my results below. My question is, under different parameters, I sometimes get values like "Variable usage = 115.35%" or "Variable usage = 0%," which is confusing to me. Since my data is generated through formulas for simulation purposes, there should be clear causal relationships. I'm not sure why this is happening, and I hope you can help clarify. Best regards!

```python
# Load time series data
file_path = "new_data.txt"
df = pd.read_csv(
    file_path,
    sep='\s+',  # Use whitespace as separator
    header=None  # No header
)
X_np = df.values
X = torch.tensor(X_np[np.newaxis], dtype=torch.float32, device=device)

# Visualize raw time series
fig, axarr = plt.subplots(1, 2, figsize=(16, 5))
axarr[0].plot(X_np)
axarr[0].set_xlabel('T')
axarr[0].set_title('Entire time series')
axarr[1].plot(X_np[:50])
axarr[1].set_xlabel('T')
axarr[1].set_title('First 50 time points')
plt.tight_layout()
plt.show()

# Initialize cLSTM model
clstm = cLSTM(X.shape[-1], hidden=100).cuda(device=device)

# Train model using ISTA optimizer
train_loss_list = train_model_ista(
    clstm, X,
    context=20,
    lam=20,
    lam_ridge=1e-2,
    lr=1e-4,
    max_iter=5000,
    check_every=50
)

# Plot training loss curve
cpu_loss_list = [loss.cpu().numpy() for loss in train_loss_list]
plt.figure(figsize=(8, 5))
plt.plot(50 * np.arange(len(cpu_loss_list)), cpu_loss_list)
plt.title('cLSTM training')
plt.ylabel('Loss')
plt.xlabel('Training steps')
plt.tight_layout()
plt.show()

# Extract estimated Granger causality matrix
GC_est = clstm.GC().cpu().data.numpy()  # Possibly a binary matrix
gc_matrix = clstm.GC(threshold=False).cpu().detach().numpy()  # Raw values

# Softmax normalization function
def softmax(x):
    e_x = np.exp(x - np.max(x))  # Prevent numerical overflow
    return e_x / e_x.sum(axis=0)

# Apply softmax along each row
GC_softmax = np.apply_along_axis(softmax, axis=1, arr=GC_est)

# Print raw Granger causality matrix
print("Granger Causality Matrix (Raw Values):")
print(gc_matrix)

# Print estimated variable usage (GC_est is expected to be binary)
print('Estimated variable usage = %.2f%%' % (100 * np.mean(GC_est)))

# Compute global causal weights using leading eigenvector
def compute_causal_weights(matrix):
    eigenvalues, eigenvectors = np.linalg.eig(matrix.T)
    sorted_idx = np.argsort(np.abs(eigenvalues))[::-1]
    principal_vec = eigenvectors[:, sorted_idx[0]].real
    weights = principal_vec / np.sum(np.abs(principal_vec))
    return np.around(weights * 100, 2)

causal_weights = compute_causal_weights(gc_matrix)

# Print global influence weights of variables
print("\n=== Global Influence Weights of Variables ===")
for i, w in enumerate(causal_weights):
    print(f"variables x{i+1}: {w}%")

# Visualize estimated Granger causality
plt.figure(figsize=(6, 5))
plt.imshow(GC_est, cmap='Blues', vmin=0, vmax=1)
plt.title('Estimated Granger Causality')
plt.ylabel('Affected series')
plt.xlabel('Causal series')
plt.xticks([])
plt.yticks([])
plt.colorbar()
plt.tight_layout()
plt.show()

# Print column-wise softmax sums (to examine causal distribution)
column_sums = np.sum(GC_softmax, axis=0)
print(column_sums)
```
This is all the code I'm using in my notebook. I realized that I had modified the GC function definition in the clstm file so that threshold=False by default. That’s likely the reason why the values can exceed 1.

What I actually want to do is to output the influence of each variable on the others, so I defined the compute_causal_weights function for that purpose.
This function computes the global influence weight of each variable in a multivariate time series based on the leading eigenvector of the transposed Granger causality matrix.
thank you again for your answer

**Comment by reherese** [comment]
My data is 500×6, indicating 500 time steps and 6 variables with strong causal relationships. I have another question: will the accuracy of GC decrease as the amount of data decreases?

**Comment by iancovert** [comment]
Hi, this is pretty hard to read without your code in triple quotes (```). Maybe you can fix the formatting?

In the meantime, I wonder if the problem is how you defined `GC_est`. You wrote `GC_est = clstm.GC().cpu().data.numpy()`, which based on my definition [here](https://github.com/iancovert/Neural-GC/blob/master/models/clstm.py#L78) should threshold the values and ensure a binary matrix. The mean calculated in `print('Estimated variable usage = %.2f%%' % (100 * np.mean(GC_est)))` then will not exceed 1. Did you maybe change the definition of the `GC` function?

I'm not sure what you want to do with `GC_softmax`, it doesn't seem like you use that variable. I didn't read the rest of the code after this.

For your other question, yes I would expect the accuracy to reduce with decreasing data. Like any ML or statistical model, you need data to learn, and our experiments in the paper definitely showed better results with more data (see [Table 1](https://arxiv.org/pdf/1802.05842)).

**Comment by reherese** [comment]
I am very confused about why my variable usage rapidly changes from 100% to 0%. Have you encountered this situation before? Perhaps you have some suggestions to share."

![Image](https://github.com/user-attachments/assets/902790eb-8d4b-4511-a302-63c91538a88a)

**Comment by iancovert** [comment]
Your penalty strength is probably too high. Variable usage = 0% means that the first-layer weight matrix is all zeros.
